cli/azd/pkg/tools/azcli/subscriptions.go

```diff
@@ -0,0 +1,178 @@
+package azcli
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	azdinternal "github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/auth"
+	"github.com/azure/azure-dev/cli/azd/pkg/httputil"
+)
+
+// SubscriptionsService allows querying of subscriptions and tenants.
+type SubscriptionsService struct {
+	credentialProvider auth.MultiTenantCredentialProvider
+	userAgent          string
+	httpClient         httputil.HttpClient
+}
+
+func NewSubscriptionsService(
+	credentialProvider auth.MultiTenantCredentialProvider,
+	httpClient httputil.HttpClient) *SubscriptionsService {
+	return &SubscriptionsService{
+		userAgent:          azdinternal.MakeUserAgentString(""),
+		httpClient:         httpClient,
+		credentialProvider: credentialProvider,
+	}
+}
+
+func (ss *SubscriptionsService) createSubscriptionsClient(
+	ctx context.Context, tenantId string) (*armsubscriptions.Client, error) {
+	options := clientOptionsBuilder(ss.httpClient, ss.userAgent).BuildArmClientOptions()
+	cred, err := ss.credentialProvider.GetTokenCredential(ctx, tenantId)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := armsubscriptions.NewClient(cred, options)
```

Comment:
there's a big perf-impact from creating sdk-clients on demand.

The sdk client are expected to be created and re-used it as many times as possible for the best results